File: android/app/src/test/java/com/easybc/planner/diagnostics/DeveloperLogTest.kt
package com.easybc.planner.diagnostics

import org.junit.Assert.assertEquals
import org.junit.Test

class DeveloperLogTest {
    @Test
    fun formatIsStableAndReadable() {
        val formatted = formatDeveloperLogEntries(
            listOf(
                DeveloperLogEntry(
                    timestamp = "2026-07-12T22:00:00Z",
                    area = "migration",
                    event = "control-read-failed",
                    details = linkedMapOf("datasetId" to "primary", "error" to "not found"),
                ),
            ),
        )

        assertEquals(
            "2026-07-12T22:00:00Z [migration] control-read-failed datasetId=primary error=not found",
            formatted,
        )
    }

    @Test
    fun emptyLogHasUsefulCopy() {
        assertEquals("No diagnostic events recorded.", formatDeveloperLogEntries(emptyList()))
    }
}

File: web/src/diagnostics/developerLog.test.ts
import { describe, expect, it } from "vitest";
import { diagnosticDetails, formatDeveloperLog } from "./developerLog";

describe("developer diagnostics", () => {
  it("formats entries as portable plain text", () => {
    expect(
      formatDeveloperLog([
        {
          timestamp: "2026-07-12T22:00:00Z",
          area: "migration",
          event: "control-read-failed",
          details: { datasetId: "primary", error: "not found" },
        },
      ]),
    ).toBe(
      "2026-07-12T22:00:00Z [migration] control-read-failed datasetId=primary error=not found",
    );
  });

  it("normalizes multiline details and caps their length", () => {
    const details = diagnosticDetails({ error: `first\nsecond ${"x".repeat(600)}` });
    expect(details.error).not.toContain("\n");
    expect(details.error).toHaveLength(500);
  });
});
